breakdown/mem0.md

```diff
@@ -0,0 +1,307 @@
+---
+title: "Memo & Memo-Graph: A Technical Breakdown of Scalable Long-Term Memory for AI Agents"
+short_title: Memo & Memo-Graph
+description: "Technical analysis of Memo, a scalable memory architecture for LLMs, and its graph-based variant, Memo-Graph, designed for long-term conversational coherence."
+date: 2025-07-30
+authors:
+  - 0xm
+tags:
+  - breakdown
+  - deepdive
+  - memory
+  - llm
+  - agent
+  - architecture
+  - graph-database
+toc: true
+---
+
+## 1. Overview
+
+**Mem0** is a memory-centric architecture designed to overcome the fixed-context window limitations of Large Language Models (LLMs), enabling consistent, long-term, multi-session conversations. It extracts, consolidates, and retrieves salient information dynamically.
+
+**Mem0-Graph** (MemO⁸) extends Memo by representing memories as a knowledge graph, enabling reasoning over complex relational structures between conversational elements.
+
+## Core Technical Approaches and Innovations
+
+- **Dynamic Memory Pipeline:** Two-phase process—Extraction (identifies important facts) and Update (integrates facts into persistent memory).
+- **LLM-Powered Memory Operations:** Uses LLM function-calling to decide ADD, UPDATE, DELETE, or NOOP for memory management.
+- **Graph-Based Memory Representation (Memo-Graph):** Models information as a directed labeled graph (entities as nodes, relationships as edges).
+- **Dual-Contextual Prompting:** Extraction uses both a long-term summary and recent message history for context.
+- **Efficiency and Scalability:** Stores concise facts, reducing token consumption and latency compared to full-context or RAG approaches.
+
+## Key Architectural Components
+
+### Memo Architecture
+
+- **Extraction Phase:** LLM processes new messages, summary, and history to extract salient memory facts.
+- **Update Phase:** Retrieves similar memories and uses LLM tool call to manage memory store (ADD/UPDATE/DELETE/NOOP).
+- **Vector Database:** Stores memory embeddings for semantic similarity search.
+
+### Memo-Graph Architecture
+
+- **Entity Extractor:** LLM identifies key entities.
+- **Relations Generator:** LLM defines relationships, creating triplets (e.g., (Alice, LIVES_IN, San Francisco)).
+- **Conflict Detector & Update Resolver:** LLM resolves contradictions, marking outdated relationships as invalid.
+- **Graph Database (Neo4j):** Stores nodes and edges of the memory graph.
+
+## Visual Illustration
+
+### Memo System Architecture
```

Comment:
In the next part of the text, there’s a section about System Architecture. This might confuse readers. Should we update the outline to make it clearer? Maybe do a merge?